hanadb_exporter/exporters/prometheus_exporter.py

```diff
@@ -31,48 +32,77 @@ class SapHanaCollector(object):
     SAP HANA database data exporter
     """
 
-    def __init__(self, conector):
+    def __init__(self, connector):
         super(SapHanaCollector, self).__init__()
         self._logger = logging.getLogger(__name__)
-        self._hdb_connector = conector
+        self._hdb_connector = connector
 
-    def _execute(self, metric):
+    def _execute(self, query):
         """
-        Create metric object
+        Execute query and return a QueryResult object
 
         Args:
-            metric (dict): query, info, type structure dictionary
+            query (string): SQL query
         """
         try:
-            value = self._hdb_connector.query(metric['query'])
-
-            if metric['type'] == core.GaugeMetricFamily:
-                metric_obj = self._manage_gauge(metric, value)
-            else:
-                raise NotImplementedError('{} type not implemented'.format(metric['type']))
-
-            return metric_obj
+            query_result = self._hdb_connector.query(query)
+            return query_result
         except KeyError as err:
             raise MalformedMetric(err)
 
-    def _manage_gauge(self, metric, value):
+    def _format_query_result(self, query_result):
+        """
+        Format query results to match column names with their values for each row
+        Returns nested list, containing tuples (column_name, value)
+
+        Args:
+            query_result (obj): QueryResult object
+        """
+        query_columns = []
+        formatted_query_result = []
+        for meta in query_result.metadata:
+            query_columns.append(meta[0])
+        for record in query_result.records:
+            formatted_query_result.append(list(itertools.izip(query_columns, record)))
+        return formatted_query_result
+
+    def _manage_gauge(self, metric, formatted_query_result):
         """
         Manage Gauge type metric
+
+        Args:
+            metric (dict): a dictionary containing information about the metric
+            formatted_query_result (nested list): query formated by _format_query_result method
         """
-        # Label not set
-        metric_obj = core.GaugeMetricFamily(*metric['info'])
-        if not metric['info'][3]:
-            metric_obj.add_metric([], str(value[0][0]))
-        else:
-            for label_item in value:
-                self._logger.info('%s: %s' % (label_item[0], label_item[1]))
-                metric_obj.add_metric([label_item[0]], float(label_item[1]))
+        metric_obj = core.GaugeMetricFamily(metric['name'], metric['description'], None, metric['labels'], metric['unit'])
+        for row in formatted_query_result:
+            labels = []
+            value = 0
```

Comment:
I should avoid this `0`. If we don't have any value, we should raise an exception and don't create the metric (or create the metric with an invalid value)

Reply:
0 would be the default value if we don't pass any value in the JSON file, plus sometimes we are not interested in the value itself but the labels, therefor we need to pass a value (convention), a metric with a 0 value and valid labels will be a valid metric and there is nothing wrong with that.

Reply:
@ayoub-belarbi be careful because it could create the following misleading effect: 
- CPU_SYSTEM_TIME = Null on the query result (because it could not be collected)
- hanadb_cpu_system_time = 0, that could be misleading as the would state 0% of cpu usage, when we don't know.

I would log an error an error in case the the value is None saying that the metric could not be collected. If the value can really be ignored, I would let the Query to treat that (like, returning 0 if the CPU field is empty).

Reply:
@diegoakechi Very valid point with your example!